driver/pace/driver/run.py

```diff
@@ -292,34 +394,50 @@ def from_dict(cls, kwargs: Dict[str, Any]) -> "DriverConfig":
             data=kwargs.get("initialization_config", {}),
             config=dacite.Config(strict=True),
         )
+
         for derived_name in ("dt_atmos", "layout", "npx", "npy", "npz", "ntiles"):
             if derived_name in kwargs["dycore_config"]:
                 raise ValueError(
                     f"you cannot set {derived_name} directly in dycore_config, "
                     "as it is determined based on top-level configuration"
                 )
-        kwargs["dycore_config"]["layout"] = kwargs["layout"]
-        kwargs["dycore_config"]["dt_atmos"] = kwargs["dt_atmos"]
-        kwargs["dycore_config"]["npx"] = kwargs["nx_tile"] + 1
-        kwargs["dycore_config"]["npy"] = kwargs["nx_tile"] + 1
-        kwargs["dycore_config"]["npz"] = kwargs["nz"]
-        kwargs["dycore_config"]["ntiles"] = 6
-        for derived_name in ("dt_atmos", "layout", "npx", "npy", "npz"):
-            if derived_name in kwargs["physics_config"]:
-                raise ValueError(
-                    f"you cannot set {derived_name} directly in physics_config, "
-                    "as it is determined based on top-level configuration"
-                )
-        kwargs["physics_config"]["layout"] = kwargs["layout"]
-        kwargs["physics_config"]["dt_atmos"] = kwargs["dt_atmos"]
-        kwargs["physics_config"]["npx"] = kwargs["nx_tile"] + 1
-        kwargs["physics_config"]["npy"] = kwargs["nx_tile"] + 1
-        kwargs["physics_config"]["npz"] = kwargs["nz"]
+
+        kwargs["dycore_config"] = dacite.from_dict(
+            data_class=fv3core.DynamicalCoreConfig,
+            data=kwargs.get("dycore_config", {}),
+            config=dacite.Config(strict=True),
+        )
+        kwargs["physics_config"] = dacite.from_dict(
+            data_class=fv3gfs.physics.PhysicsConfig,
+            data=kwargs.get("physics_config", {}),
+            config=dacite.Config(strict=True),
+        )
+
+        if initialization_type != "serialbox":
+            kwargs["layout"] = tuple(kwargs["layout"])
+            kwargs["dycore_config"].layout = kwargs["layout"]
+            kwargs["dycore_config"].dt_atmos = kwargs["dt_atmos"]
+            kwargs["dycore_config"].npx = kwargs["nx_tile"] + 1
+            kwargs["dycore_config"].npy = kwargs["nx_tile"] + 1
+            kwargs["dycore_config"].npz = kwargs["nz"]
+            kwargs["dycore_config"].ntiles = 6
+            kwargs["physics_config"].layout = kwargs["layout"]
+            kwargs["physics_config"].dt_atmos = kwargs["dt_atmos"]
+            kwargs["physics_config"].npx = kwargs["nx_tile"] + 1
+            kwargs["physics_config"].npy = kwargs["nx_tile"] + 1
+            kwargs["physics_config"].npz = kwargs["nz"]
+
+        elif initialization_type == "serialbox":
+            kwargs["nx_tile"] = kwargs["dycore_config"].npx - 1
+            kwargs["nz"] = kwargs["dycore_config"].npz
+            kwargs["layout"] = tuple(kwargs["dycore_config"].layout)
```

Comment:
This still makes me nervous (changing the behavior of the coupling when initialization_type="serialbox"), a consequence of that choice is this has pretty odd behaviors in the case where initialization_type="serialbox" and namelist_override is not used for one or both of dycore_config and physics_config. We don't plan to use it but it's a valid configuration option presented to the user.

Is it that bad to rely on the yaml for the top-level settings (nx_tile, nz, layout) and expect the user to make sure these correspond with the serialbox directory?

Admittedly this is a lot cleaner and less dangerous than before, so I'll leave it up to you. If you keep it this way, I would suggest adding a check that raises an exception if initialization_type="serialbox" and namelist_override is not given for either of dycore_config and physics_config.

Reply:
You can actually still have serialbox initialization and pass in dycore_config and physics_config. I think we don't expect people to do that (assuming if you have serialized data, you ran fortran with a nml file), but in theory, it is allowed and shouldn't break the code.

Reply:
What I mean is, the normal expectation here is that you don't pass `npx` in `dycore_config`, and instead you pass `nx_tile`. If you take a configuration file that does this and modify the initailization_type, npx will be set to 0 and no exception will be raised. This could be pretty hard to detect and debug.

Reply:
Overriding the owning class behavior based on the choice of subclass for initialization makes me nervous because of how easy it is to not notice potential bugs like that, and how hard it is to detect them when they do happen.

If we consider the "serialbox" initialization to be a use-at-your-own-risk configuration that may cause strange bugs then this is OK, but I'm not sure the tradeoff of not copying these variables from the namelist to the yaml outweighs the risk of us running into one of these bugs. We can merge this as-is if you think it will be fine.

Reply:
Actually, is there a more important bug here? Does `dt_atmos` as well as `days`, `hours`, `minutes`, `seconds` need to be copied from DycoreConfig to DriverConfig as well?